Add blog post about changes to `UnitSystem` class and their implications (#2618)
home-assistant/developers.home-assistant: blog/2025-03-24-frozen-unit-system.md

> ---

Reviewer: Please change the name of the file to today's date. We already published a blog yesterday.

author: Shay Levy
authorURL: https://github.com/thecode

> title: "Changes to UnitSystem class"

Reviewer 2: ```suggestion
title: "UnitSystem dataclass is now frozen"
```

---

> `UnitSystem` class is changed to a frozen data class, all classes derived from it are now frozen.

Reviewer: ```suggestion
`UnitSystem` class is changed to a frozen data class, and all intances derived from it are now frozen.
```

> `UnitSystem` class is changed to a frozen data class, and all intances derived from it are now frozen.

Reviewer: ```suggestion
`UnitSystem` class is changed to a frozen data class, and all instances derived from it are now frozen.
```

Reviewer: Sorry about the typo.

The following unit systems that derived from the `UnitSystem` class are now frozen:

- `METRIC_SYSTEM`
- `US_CUSTOMARY_SYSTEM`

> The reason for this change is that Unit systems are constants that should not be modified. An integration that modify these constants can break unit conversions and create undesired output for other components of Home Assistant.

Reviewer 2: ```suggestion
The reason for this change is that Unit systems are constants that should not be modified. An integration that modifies these constants can break unit conversions and create undesired output for other components of Home Assistant.
```

With a frozen data class an attempt to modify the UnitSystem constant will fail:

```python
dataclasses.FrozenInstanceError: cannot assign to field 'temperature_unit'
```

> The change is introduced in the [home assistant core PR #140954](https://github.com/home-assistant/core/pull/140954).

Reviewer 2: ```suggestion
This change was introduced in the [home assistant core PR #140954](https://github.com/home-assistant/core/pull/140954).
```
